rpc/http.go

```diff
@@ -313,3 +97,99 @@ func (client *httpClient) Close() {
 func (client *httpClient) SupportedModules() (map[string]string, error) {
 	return SupportedModules(client)
 }
+
+// httpStream reads and writes HTTP requests and responses and can be used
+// as the ReadWriteCloser for the ServerCodec.
+type httpStream struct {
+	mu              sync.Mutex
+	request         *http.Request
+	firstRead     bool
+	response        http.ResponseWriter
+	responseWritten chan struct{}
+}
+
+// Read the body of the next HTTP request.
+func (rw *httpStream) Read(p []byte) (int, error) {
+	n, err := rw.request.Body.Read(p)
+	if n == 0 && err == io.EOF && !rw.firstRead { // request already read
+		<-rw.responseWritten // wait for response to be written (http is sync)
+		return 0, io.EOF     // indicate that the request has been read
+	}
+	rw.firstRead = false
+	return n, err
+}
```

Comment:
I'm not sure it's safe to make `Read` and `Write` a one shot call. You really have no guarantee that all the data will be returned to you in one go (e.g. the input is larger that some Go internal buffer over which we don't have control).